Default and expanded errors for rustc

jonathandturner opened the pull request:
Proposed update to the Rust compiler error output format.

[rendered](https://github.com/jonathandturner/rfcs/blob/master/text/0000-default-and-expanded-rustc-errors.md) (last updated: 6/27/2016 8:30am)

AndrewBrinker commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,191 @@
+
+There are a few unresolved questions:
+* Editors that rely on pattern-matching the compiler output will need to be updated.  It's an open question how best to transition to using the new errors.  There is on-going discussion of standardizing the JSON output, which could also be used.
+* Can additional error notes be shown without the "rainbow problem" where too many colors and too much boldness cause errors to beocome less readable?
```
"beocome" -> "become"

ticki replied:
I am sure this is a subjective problem. Some people prefer the colorful approach, others not so much, so I think configurability is the key here.

petrochenkov commented:
I'm probably not the target auditory, but here's one data point.
The green areas is where my sight is usually focused when I look at error messages. The time frame is about few seconds. The rest of the area, especially long explanations, is pretty much noise and is never read.
![error_message](https://cloud.githubusercontent.com/assets/5751617/15876276/32b80414-2d15-11e6-972a-6df6b25dbd7f.png)

jonathandturner commented:
@petrochenkov - thanks!  That's actually a really cool analysis.  It's totally natural to skip around and only read the parts you want to read.

I don't talk about it much in this RFC, but I did a quick survey of how small number of expert Rust developers used the current error system.  Many of them just looked at the line number (and possibly the start of the error message) and ignored the rest.  They trusted their ability to debug with just the location in mind.

Being able to be visually parsed as you described is definitely an improvement and to have that much coverage in a few seconds is very promising.

Thanks again.

bstrie commented:
I feel like there are two orthogonal RFCs here. We should be able to discuss the new error formatting independently of whether explain should be redesigned.

phil-opp commented:
IMO the separator of the line number column is a bit noisy. How about just `|` instead of `|>`?:

<img width="350" alt="rust-new-error-format-" src="https://cloud.githubusercontent.com/assets/1131315/15889231/998500d4-2d6a-11e6-8346-b5447881e20f.png">

I think it's less visual clutter and much easier to read, especially the last lines. For example, there is no `|> }` anymore.

ticki commented:
I agree with @petrochenkov. A lot of it is not relevant at least 90% of the time, but it may be so for new beginners, so having different log-level-esque errors would be very cool.

ticki commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,191 @@
+
+# Alternatives
+
+Rather than using the proposed error format format, we could only provide the verbose --explain style that is proposed in this RFC.  Famous programmers like [John Carmack](https://twitter.com/ID_AA_Carmack/status/735197548034412546) have praised the Elm error format.
```
[dybuk](https://github.com/ticki/dybuk) is an implementation of something like this.

ticki commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,191 @@
+
+Rather than using the proposed error format format, we could only provide the verbose --explain style that is proposed in this RFC.  Famous programmers like [John Carmack](https://twitter.com/ID_AA_Carmack/status/735197548034412546) have praised the Elm error format.
+
+![Image of Elm error](http://www.jonathanturner.org/images/elm_error.jpg)
```
This image is rather visual trickery. You should make sure that the same colorscheme is used throughout the RFC, to avoid unintended bias.

jonathandturner commented:
@bstrie - possibly, though this lays out what should be default as well (vs say just making the Elm-style the only error output)

alexcrichton commented:
I'm pretty excited to see this move forward, I've been using these errors messages on nightly and they're *fantastic* compared to the old system, and I can't wait to basically just get more of them.

I also really like the idea of `--explain` going inline with your text and being more descriptive of the code you literally wrote, that seems to be invaluable for learning code quickly as it reduces the path to working code (you just copy/paste what the errors tell you to do) and may help understanding of the rules in abstract over time. Maybe even one day we can have a mode of the compiler that auto-applies all these fancy suggestions and writes your code for you!

One part about removing the error code of `--explain` though is that we're losing a way to uniquely identify errors. I don't think this is too useful on the command line (kinda annoying to copy/paste), but I've heard it's very useful for *googling* or search around for solutions others have applied (e.g. when we don't have a suggestion to apply). I wonder if `--explain` could just expand the error text with a code somewhere in it though? Or maybe we could just unconditionally put the error codes in there all the time somewhere inconspicous (they're pretty small). 

My only other thought is that on the removal of "error: aborting due to 2 previous errors" I actually really like seeing the total number of errors that were emitted sometimes. If you do some invasive refactoring and fix everywhere you remembered, this gives a great "progress bar" of how far you have left to go. Maybe that cold be folded into some message elsewhere though? I do agree that the dedicated "error: " line is a bit overkill so I don't really have a preference as to *where* it shows up so long as it's at the end.

AndrewBrinker commented:
Piggybacking on @alexcrichton's point, I know I've definitely appreciated the searchable error codes before. It makes it a lot easier to find explanations or people with similar situations, and a lot easier to help other people (they can say "I got this error code" and you have a solid sense of what went wrong, very quickly and succinctly).

jonathandturner commented:
@alexcrichton @AndrewBrinker - Definitely agree that keeping it google-able is a good design feature.  Rather than using the --explain EXXX pattern, we could just have the error code be part of the error message.  Maybe something like:

<img width="482" alt="screen shot 2016-06-08 at 10 07 13 am" src="https://cloud.githubusercontent.com/assets/547158/15897045/d3684022-2d60-11e6-9c1d-31d5b2f9f55a.png">

nikomatsakis commented:
I think just keeping the codes, without having them be particularly important to explain, makes a lot of sense.

nikomatsakis commented:
(Oh, and I see no reason we can't also tell you the total number of errors while we let you know about `--explain`)

nikomatsakis commented:
@bstrie Do you object to the idea of having `--explain` incorporate parts of the error? It feels like it fits thematically into this RFC -- which is to generally update the "look and feel" over error handling -- and I don't see why we would split it off absent some significant controversy.

For example, @jonathandturner and I spent a while debating whether `--explain` ought to be the *default* -- or even if it could be made into the only option! -- which suggests that the two questions (default output, explain output) are quite linked. (We ultimately decided against that, obviously.)

alexcrichton commented:
@jonathandturner [that](https://github.com/rust-lang/rfcs/pull/1644#issuecomment-224600271) looks really good to me! We could maybe toy around with other concepts like:

```
error[E0455]: cannot borrow `v` as mutable more than once at a time
```

but that's all in the weeds of bike shedding :). My basic point was that we probably shouldn't jettison error codes even if `--explain` doesn't take them, and it sounds like we're all on board with that!

@nikomatsakis yeah I also agree it should be trivial to keep the number of errors while we also explain `--explain` (see what I did there?).

ticki commented:
@camlorn I will transcribe them for you, two seconds.

ticki commented:
@camlorn The first image is the old error format, where the path is written in the start of every line.

The second image is the new format:

```
error: `self` is not available in a static method. Maybe a `self` argument is missing? [--explain E0424]
  --> vec.rs:94:23
   |>
94 |>         debug_assert!(self.len <= block.size() / mem::size_of::<T>(), "Block not large enough to \
   |>                       ^^^^
```

The third example describes the long form:

```
error: cannot move out of borrowed content:
  --> /Users/jturner/Source/errors/borrowck-move-out-of-vec-tail.rs:30:37

I'm trying to track the ownership of the contents of `tail`, which is
borrowed, through this match statement:

29 |>            match tail {

In this match, you use a pattern which will bind to a member of `tail` as a
value. This will move the contents out of `tail`. Because `tail` is
borrowed, you can't safely move the contents:

30 |>              [Foo { string: aa },
   |>                             ^^ cannot move out of borrowed content

You can avoid moving the contents out by using a reference in the pattern, rather
than a move. A naive fix might look like:

30 |>              [Foo { string: ref aa },
```

The fourth image is the same as the first one. The 6th image is the same as the first one, but with the header (i.e., the error) highlighted. The 7th is the same image again with the line numbers highlighted. The 8th is the same image again with the code span highlighted.

The 9th image is:

```
error: `Shrinkable` is not a trait [--explain E0404]
 --> rust_error2.rs:5:6
  |>
3 |> type Shrinkable = Growable;
  |> --------------------------- type aliases cannot be used for traits
5 |> impl Shrinkable for i32 { }
  |>      ^^^^^^^^^^ `Shrinkable` is not a trait
```

The 10th image is the same as the third one. The 11th example is another example of the new format.

The last image is Elm's format:

```
Detected errors in 1 module.

-- ALIAS PROBLEM ---------------------------------------------- ././Maze.elm

This type alias is recursive, forming an infinite type!

21|>type alias Node =
22|>    { x : Int
23|>    , y : Int
24|>    , children : List Node
25|>    }

When I expand a recursive type alias, it just keeps getting bigger and bigger.
So dealiasing results in an infinitely large type! Try this instead:

    type Node
        = Node { x : Int, y : Int, children : List Node }

This is kind of a subtle distinction. I suggest the naive fix, but you can
often do something a bit nicer. So I would recommend reading more at:
<https://github.com/elm-lang/elm-compiler/blob/0.17.0/hints/recursive-alias.md>
```

That's all.

nikomatsakis commented:
Interesting. There has also been a discussion around a so-called "traditional" format which would (I think) be something highly condensed of the form

```
file:line: error: message
```

and maybe even no additional detail? This would be useful for older
tools that don't really know anything about Rust; it might also help
for screen reader support, I suppose, though you would probably be
interested in having the full details I imagine (i.e., the notes where
a borrow begins and so forth)

On Wed, Jun 08, 2016 at 12:14:45PM -0700, Austin Hicks wrote:
> Can we please get these as text?  I don't know what's being proposed because the RFC only uses images, and images don't read with screen readers.
> 
> Will this make errors shorter?  I have to scroll the console somewhat often even with 4 or 5 errors.  I'd love a nice one or two line output format, but that's probably not what this is.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-lang/rfcs/pull/1644#issuecomment-224697587

ticki commented:
I wonder what we can do for people like @camlorn with respect to the error format. @camlorn, what would be your ideal error format?

camlorn commented:
thanks for the transcription.

As for what we can do generally?  Convince Microsoft to let me set the font in the console even smaller.  I don't need readable...

More seriously...

This is an opinion.  I'm kind of unique in a lot of ways, not least that I'm one of the few blind people who considers C++ easy.  My considerations may not be the considerations of a new blind programmer.

Firstly, we don't need `--explain` for advanced rust people.  This takes terminal space.  I know it exists.  Reminding me of it is just an annoying thing to read.  Obviously it needs to be on by default, but turning it off somehow would be nice.  We could move the error code to be in brackets after the line numbers, or just in brackets after the message.

Prefixing with `|>` is pretty much the same as prefixing with `///`.  All my comments on the "Only allow /// for docs" rfc apply here: it's annoying but readable if I have to.

The underlines to show where the error is are useless completely.  I don't think there's a particularly good solution unless we have a character which will be always reserved.  If we did, we could surround the error with it in the same manner that you use parens.  As it stands, I'd like to murder the underlines in the most painful way possible and not only for Rust.  Surrounding the error's source with something might actually be helpful.

I've never quite used the information on borrow scopes.  I don't know why.  I think that part of it is that reading code in the terminal is inconvenient.  I also find that my C++ experience puts me in a good place to reason about lifetimes.

Showing the surrounding code is not helpful.  If I want to see the surrounding code, I'll use go to line in my editor.  I think this one is a fundamental difference between blind and sighted people.  There is no way for me to glance at anything, so the cost of reading the code in the terminal and then reading the code in the editor is about twice that of just using the editor.  If I have to scroll to see it, it's more.  If we can find a way to emphasize the error, showing the code for the error might be useful.

If I had to make a suggestion, provide the line numbers of the borrow regions, possibly with the text of the line.  This might be `{` and `}` in practice, so that might not be so useful.

If we assume that this is a hypothetical terse error mode that you explicitly enable with an environment variable, you could have a switch to explicitly disable it.  In that case, if I need the information in the expanded version, I can still get it.

Sorry I can't speak more definitely.  I started a large project in Rust, but the larger project in C++ took priority for now.  This means that I don't have enough experience to have a workflow that's a procedure as such.  With the C++ compiler, I could document exactly what I press because it's always the same.  With Rust, it's not quite to the level of mechanical procedure yet.

brson commented:
I like the direction of this RFC generally.

`--explain` is a stable compiler interface. How can we change it? I agree that `--explain` is a sweet name for this functionality, but I think we need to use a different one.

It's not clear to me what the fate of error codes is here, but my impression is that you are proposing to remove them, but your images contain the current `--explain E0002`. Can you make this more clear, and if you are removing error codes can remove them from the graphics?

> "Famous programmers like John Carmack have praised the Elm error format."

John Carmack's fame doesn't really have any bearing on the validity of the proposal.

Also, I don't think RFCs should be giving props to other contributors in the summary. When creating the RFC process we made an explicit decision not to include author information, to avoid making RFC authorship itself an incentive (I think). *If* authorship is something that should be included in RFCs then the process itself should be modified to include that metadata, including crediting the actual author, not just their collaborators. If giving credit in the RFC text is something we want to encourage, then I don't know that the summary is the place for it. The summary is for summarizing the content of the RFC.

nagisa commented:
At the very least we should have a fully working json error output, which one could pass through an arbitrary filter of the form `rustc >/dev/null 2>&1 | awesome_error_printer` and thus making it possible to make rust output accessible, even if not by default. Sure that doesn’t change the fact about default experience being junk for people with special needs, but at least there would be an option, however minuscule.

ticki commented:
@camlorn @nagisa  We should have a `ACESSIBLITY_MODE` variable.

nielsle commented:
I think that it would be helpful for newbies to spend an  entire sentence to introduce `--explain  E040` 
~~~
error: `Shrinkable` is not a trait [E040]
--> rust_error2.rs:5:6
  |>
3 |> type Shrinkable = Growable;
  |> --------------------------- type aliases cannot be used for traits
5 |> impl Shrinkable for i32 { }
  |>      ^^^^^^^^^^ `Shrinkable` is not a trait
For a detailed description of the error  type: rustc --explain E040
~~~
It would be great  to have an  `ACESSIBILITY_MODE`. Perhaps it could also control warnings about unoptimized builds #967.

camlorn commented:
What solves the problem generally is the ability to just write custom error templates.  My accessible mode is probably not the next person's accessible mode, and I think that more blind people need to be programming in Rust before we can go there.  Unfortunately, I don't think this is likely to happen anytime soon.  I might be the only one.

I don't like the idea of piping through a script. Json error output could be useful for other reasons.  I suppose that it can somehow be hacked together, but Windows scripting support is really not so great.  And also whatever you write will have to deal with cargo output too.

The easiest way to make this immediately better is to let me configure the number of errors to show.  If you set it to 2 instead of 5, then the terminal doesn't scroll.  I'm going to also make a point of asking the head dev of my screen reader if automatic scrolling is possible.  This doesn't fix the issue of a lot of noise and extra lines that I have to arrow past, but it does fix needing to press a ton of extra stuff to just move the console up a page.  This also looks like it might be a bit shorter, so it's possible it'll fix that part of it on it's own.

I'm not really concerned about it.  It could be better.  But Rust is far from the top of the list of inaccessible things that make my life with technology harder.

brson commented:
My comments could have been more constructive. Let me respond to myself.

>     "Famous programmers like John Carmack have praised the Elm error format."
> 
> John Carmack's fame doesn't really have any bearing on the validity of the proposal.

Maybe just change 'fame' to something more relevant, like 'well-respected', something that better indicates why his opinion is important.

> Also, I don't think RFCs should be giving props to other contributors in the summary. When creating the RFC process we made an explicit decision not to include author information, to avoid making RFC authorship itself an incentive (I think). If authorship is something that should be included in RFCs then the process itself should be modified to include that metadata, including crediting the actual author, not just their collaborators. If giving credit in the RFC text is something we want to encourage, then I don't know that the summary is the place for it. The summary is for summarizing the content of the RFC.

I phrased this too strongly. I just had a sudden observation that we've developed ad-hoc ways to credit people that doesn't fit into the process as implemented. I should have phrased this more like 'maybe we should reconsider how to give credit in RFCs'. RFC process is organic so it may be perfectly fine to do it like this (author credit in the commit log, additional credits in the summary), but it is awkward.

jonathandturner commented:
@camlorn - I'm currently working on a refactor of error reporting into its own crate that should make it easier to configure your error messages.  I'd planned to support a 'default' (the proposed error with underlines), 'expanded' (the proposed new --explain), and 'json'.  Once the refactor is in place it should be straightforward to add a format that's more friendly to screen readers.

Thanks for your detailed description of a screen reader-friendly error format.  I suspect you would not be the only person to use it, and that other people would also find it useful.  To make sure I understand, would the format be something like this in practice?  

```
error[E0499]: cannot borrow `*x` as mutable more than once at a time at src/test/compile-fail/borrowck/borrowck-borrow-mut-object-twice.rs:21:5
error: "second mutable borrow occurs here" at 21:4
note: "first mutable borrow occurs here" at 20:13
note: "first bottow ends here" at 22:1
```

I'd love to brainstorm/iterate on putting together a more compact, reader-friendly format.

ticki commented:
@camlorn @jonathandturner I am currently working on a "demand-driven" error formatter for blind people. A session would look like this:

```
Error in line 245
show
cannot move out of borrowed context.
code
      self.move()
next
Error in line 247
show
cannot borrow `*x` as mutable more than once at a time
recompile
```

I would like your input on this @camlorn.

I try to avoid sign clutter, so there's no e.g. `$` in the command prompt, possibly making it more screen-reader friendly.

bjz commented:
I would thumbs-up the new error format, but the fact that it has been linked with the `--explain` suggestion prevents me from doing so. I've been using the new error messages by default this past month and find it very jarring to go back! I'm very happy with the work that has been done!

Echoing @alexcrichton - I would love to see the total error/warning counts restored - I find them a nice way to see your progress when working through a refactor.

nikomatsakis commented:
@bjz 

> I would thumbs-up the new error format, but the fact that it has been linked with the --explain suggestion prevents me from doing so.

Can you expand a bit here? Are you saying that you think it's better to have `--explain` continue to be "canned" text, withholding judgement until you see something more concrete, or something else entirely?

bjz commented:
Yes. I can confidently thumbs up the new error format because I've actually used it, and the experience has been very positive. The thinking behind it has been tested, and experimented with, and the way it was introduced.

I like the general idea of the `--explain` part of the RFC, but as a UI suggestion I feel it needs some practical user testing and experimentation in a separate forum before we have the same level of confidence as with the new error formatting.

camlorn commented:
I got dragged into other stuff, thus the sudden silence.

@jonathandturner 

I doubt anyone blind would install a third-party error formatter unless they already knew about it, and there's no way for it to really percolate.  There used to be communities of blind programmers, but they all sort of disappeared somewhere along the way (now that I think about it, this would have been around the time that IDEs became significantly less accessible, almost overnight).  The largest I'm now aware of is a group of primarily teenagers making something called audiogames and using a closed source and buggy language called BGT.  In general, the blind communities are very fragmented and it's very hard for something like this to travel across them.

If there's serious interest in doing this, it needs to be in core.  It also probably needs to look for a running screen reader (i.e. the screen reader flag on Windows) and offer (once!) to turn on the accessibility mode.  Failing that, I'm really not sure how anyone will find it.

At the very least, it would need to not require nightly.  I suppose this can be done with a Cargo command, but am not sure how you'd get Cargo to use it.

@ticki 

That's an interesting format, but misses one of the biggest problems.  Every line is expensive.  When you split it up so that the error takes `n` lines, I have to use a minimum of `n` keystrokes to get past it.

Punctuation in itself isn't actually a problem.  You pretty much have to get used to it one way or another.  The issue is things like the underlines that you see in Rust and Clang where it tries to visually show what the error is.  These take up a line and provide absolutely no information because screen readers aren't capable of showing me what they actually underline.  Admittedly this is probably scriptable on the screen reader side, but it also probably has to be done per compiler and I've never cared about them enough to bother finding out how involved it would be.

Condense it like this, and you start to get somewhere:

```
my_file.rs:245:12: cannot move out of borrowed context.
      self.move()
my_file.rs:247:15 cannot borrow `*x` as mutable more than once at a time
2 Errors.
```

Screen readers can indicate indentation if it's the first thing on the line, so grouping by indenting works for some of us.  This gets into opinionated territory again, however.  Some blind people (me) turn on indentation because it actually provides useful information.  Others (most people I know) hate it like the plague.  Some people even refuse to use Python because of it.

I'd still like to see something taking the place of the underlines here, but am not sure what.  Code is costing a line without providing much in the way of information again.

It's also worth noting that a good IDE (i.e. Microsoft making VS work with [NVDA](http://www.nvaccess.org/), then Rust integrating with VS) solves this problem even better.  In the best possible world, you hit ctrl+b or somesuch, the error list pops up, you move between them with one keystroke, and you press enter to jump directly to the file and the line.  If I thought it would do any good, my suggestion to this community would be to take the energy it takes to develop and maintain a custom error format and spend it convincing Microsoft and other IDE-making companies that accessibility is important.  A good IDE allows for the maximum possible efficiency, in theory.  But there aren't enough blind programmers that the big IDE people are taking notice of us, and it's not like this thread has enough clout behind it that they will because I called them out here.  nevertheless, this remains the best theoretical solution.

But generally, and to everyone following this little subthread, there's two things:

- I expect the total number of blind people who know about Rust and have tried it for anything beyond hello world to remain in the double digits for at least 2 more years.  It's very hard to turn this intuition into concrete evidence, but suffice it to say that applying sighted demographics to blind people fails.  Blind programmers are a smaller percentage of the whole, primarily due to a lack of math and science education (or anything I would call education period, in lots and lots of places, including significant parts of the U.S.).  My Alma Mater had multiple blind people try the CS program before me and I was exceptional in their eyes from the start because I was the first to get past the most introductory classes.  This was a completely average, not notable CS program.

- The blind community frowns on accessibility modes for lots of quite good reasons generally.  I want to say no to this as a knee-jerk reaction, but am forced to admit it could be useful.  Before we seriously go down this road, we need to determine if there's enough interest to put it in core and keep it maintained for the long haul.  We also need to determine how we're going to get more useful feedback from other blind people.  In short, this is not a small or quick undertaking.

perhaps this subthread should be moved somewhere else at this point.  Also, I apologize if I'm saying very obvious things.  I'm used to having to explain myself thoroughly to sighted devs who then proceed to completely misunderstand and/or explain how I'm wrong about the software I use, though admittedly that seems less necessary here.

jonathandturner commented:
I've posted a revised version of this RFC, specifically addressing these points:

* The `|>` separator has been changed to `|`.  
* The error code has been kept for google-ability and moved to beside the "error:" indicator
* Back-compat with `--explain [error code]` is now kept
* Additional mode `--explain errors` added. This becomes the new preferred explain mode
* Fixed typos and added a few clarifications
* Warning/error counts maintained
* Moved to more text-friendly RFC

Additional formats, like those for programmers with sight-impairments, are definitely on the table but we'll move them to a separate thread so we can close on this RFC.  As @camlorn points out, this space requires more than is in scope for this RFC.  These additional formats are (or can be) compatible with this RFC.

jonathandturner commented:
Another question to answer - 

We put the new error messages behind a flag after a pretty long discussion about them (kind of an informal RFC).  They're in nightly now under "export RUST_NEW_ERROR_FORMAT=true".  Because they've been there a while, people have already had experienced using them and giving pretty positive feedback.

Usually, after the formal RFC is when features get put behind the flag, so this work is perhaps a little out of order.  Still, if we agree to this RFC, should the new error format be turned on by default or should we have a process of verifying and validating and only then make it the default?

liigo commented on text/0000-default-and-expanded-rustc-errors.md:
````diff
@@ -0,0 +1,290 @@
+
+```
+error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
+  --> src/test/compile-fail/borrowck/borrowck-borrow-from-owned-ptr.rs:29:22
````
Indent was introduced to *align* 'error:' and '-->', but we don't need it now.

```
error: msg
   --> file
```

jonathandturner replied:
Originally, the indent aligns with the column below it.  For example:

```
 --> file
  |
2 |
```
or
```
    --> file
     |
9999 |
```
IMHO, having some indent there still looks better and makes it easier to find the filename/line info, though there are other alternatives we could explore, like:

```
   --> file
     |
9999 |
```

liigo replied:
Yes, you are right.

bjz commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,290 @@
+You can avoid moving the contents out by working with each part using a reference rather than a move.  A naive fix 
+might look this:
+
+30  |>                 [Foo { string: ref aa },
```
The code in these explain errors still uses `|>`. Was that intentional?

jonathandturner replied:
Good catch!  It was not.  I'll update it.

alexcrichton commented:
Updates all look fantastic to me, thanks @jonathandturner!

> should the new error format be turned on by default or should we have a process of verifying and validating and only then make it the default?

I would personally advocate for one cycle of behind the env var we have today, and then if it all looks good ungate it after that cycle

nikomatsakis commented:
Nice work @jonathandturner on the revisions. 👍 from me.

jonathandturner commented:
Great.  Sounds like this thing is ready to go to FCP, taking @alexcrichton of being behind a flag for a cycle.

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,290 @@
+- Rust Issue: (leave this empty)
+
+# Summary
+This RFC proposes an update to error reporting in rustc.  Its focus is to change the format of Rust error messages and improve --explain capabilities to focus on the user's code.  The end goal is for errors and explain text to be more readable, more friendly to new users, while still helping Rust coders fix bugs as quickly as possible.  We expect to follow this RFC with a supplemental RFC that pr ...
```
nit: could you wrap at 80 chars please (I'm sure this was specified somewhere, but maybe it is just a de facto thing that people do?)

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
````diff
@@ -0,0 +1,290 @@
+```
+
+[as image]
+![Image of new error flow](http://www.jonathanturner.org/images/old_errors_3.png)
````
nit: alt text is incorrect (this is the old errors, not the new ones, right?). Also I don't think you need the `[as text]`/`[as image]` tags - GH renders them as text.

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,290 @@
+
+Key parts of the code are labeled with messages to help the user understand the message.  
+
+The primary label is the label associated with the main warning/error.  It explains the **what** of the compiler message.  By reading it, the user can begin to understand what the root cause of the error or warning is.  This label is colored to match the level of the message (yellow for warning, red for error) and uses the ^^^ underline.  
```
Could the error label be standard weight text rather than bold? One problem I have with the new messages is all the bold text - it makes the really important stuff (which should be bold) harder to pick out and (in my console at least) the bold text is quite a bit harder to read than standard weight.

nrc commented:
I like the error format changes a lot and approve this RFC. However, the changes to explain seem like they could do with some experimentation before they get RFC'ed. I don't think we need an RFC to do that experimentation and I think an eventual RFC would benefit from having the experimentation done first.

While we do RFC language features before implementation, it is a bit weird to have a two-part RFC with half implemented and iterated upon, and half speculative.

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,290 @@
+error: cannot move out of borrowed content
+   --> /Users/jturner/Source/errors/borrowck-move-out-of-vec-tail.rs:30:17
+
+I’m trying to track the ownership of the contents of `tail`, which is borrowed, through this match statement:
```
This first-person, conversational style drives me nuts. I find it patronising and hard to comprehend. AFAIK there is only one other language that does this and there is no evidence that it is in any way better than traditional style.

jonathandturner replied:
Thanks for the feedback.  I should have probably reiterated again in this section that this RFC's focus is the error format, not the error writing style.  The way these errors are written will be a follow-up RFC as we get more experience with what works best.

I list some evidence, by way of mentioning it being praised by noted programmers, but anecdotally, I've been watching the feedback from other programmers around the web.  They praise the elm style for being a) one of the best error message styles, b) so thorough they use it as a way to pick back up an editing session if it's cut short on a train ride, c) more new user-friendly than other error styles.

Note, too, that we're not saying that the expanded style is better for all users, but rather it's one we should support for users who wish to have more information.  Some users may enable `--explain errors` and leave it on for all projects, and that would be fine, while others may prefer more compact formats.

nrc replied:
To clarify, I didn't mean the style of integrating user code and error mesage, I meant the style of using first and second person pronouns and anthropomorphisation of the compiler, rather than a more academic tone.

jonathandturner replied:
Right, that's why I say "We expect to follow this RFC with a supplemental RFC that provides a writing style guide for error messages and explain text with a focus on readability and education." in the RFC.  The writing style will need to be a separate RFC.

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,290 @@
+
+Inspired by Dybuk and Elm, the line numbers are separated with a 'wall', a separator formed from pipe('|') characters, to clearly distinguish what is a line number from what is source at a glance.
+
+As the wall also forms a way to visually separate distinct errors, we propose extending this concept to also support span-less notes and hints.  For example:
```
I don't understand this. AIUI, the point of the code-oriented layout is that multiple notes ('sub-errors') on the same block of code are displayed together and inline. So the wall ties the errors together rather than separates them out.

In the example below, I'm not sure what is being communicated by the `=` (which, bikeshed, I'm not a fan of because it is so commonly used in code) - are the notes meant to be tied to the walled source or distinct from it? If the former then this seems to imply that the notes come logically at the end of the code, but in fact they might in general be related to an error in the middle of the displayed lines of codes.

On a purely aesthetic level, I'm not sure if `<indent>= note...` either makes the notes more distinct or more tied to the walled code (c.f., `<no indent> note...`, but (to me) it does read as more 'line-noisey'.

jonathandturner replied:
There are multiple kinds of notes, the knowledge of which I unfortunately assume in a few places.  Hints, suggestions, notes, etc can all be associated with an error message.  Notes (and the like) which do not have span information need to still be associated with the error.  They're less common now that many notes are becoming labels, but they still exist in a few cases.  Where un-spanned notes are associated with an error, I suggest printing them at the bottom of the error block.

A bit of backstory here.  In a previous iteration of the design (currently in nightly), they are left-aligned and colored.  @alexcrichton showed us some examples at one point where he was getting a lot of error messages in a row, all with unspanned notes.  The result had a negative impact on readability because it was harder to discern each separate error.  In this version of the format, we group unspanned notes and indent them so your eye can see a whole error message as a single group, by skipping the indented line and looking for where the next message starts.

nrc replied:
Ok, that makes sense for motivation. I wonder if we can do better than `=` though.

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
```diff
@@ -0,0 +1,290 @@
+The expanded error message effectively becomes a template.  The text of the template is the educational text that is explaining the message more more detail.  The template is then populated using the source lines, labels, and spans from the same compiler message that's printed in the default mode.  This lets the message writer call out each label or span as appropriate in the expanded text.
+
+It's possible to also add additional labels that aren't necessarily shown in the default error mode but would be available in the expanded error format.  This gives the explain text writer maximal flexibility without impacting the readability of the default message.  I'm currently prototyping an implementation of how this templating could work in practice.
+
```
It seems to me that there are two orthogonal features here that users might want either or both of: template explanations and inline explanations. There seems to be use cases for both (for both beginning and advanced users) so it might be nice to have access to both separately. It might be that we want to only expose one option to rustc users, but that IDEs can configure things in more detail (this might just come down to ignoring or rendering parts of the JSON errors, in practice).

nrc replied:
As I mentioned earlier, I think the `--explain` errors should be a separate RFC, but in particular it seems premature to have an RFC without the templating language or other details specified.

jonathandturner replied:
I'm not sure I follow this point.  The templated language piece seems more engineer-related rather than design related.  We could iterate the templating piece without impacting the user's experience.

nrc replied:
I guess I think of the interface we present to Rustdoc authors as a user-facing API, not just the output of Rustdoc. Therefore, there is a design question there that should be answered in an RFC.

jonathandturner replied:
The way I'm picturing this, only compiler authors who are writing error messages will see the templates.  It's not intended to be a user-facing API for users to write their own.  Similar to the existing error reporting mechanism the compiler currently uses, it would just have an additional "backend" of outputting to a template.

It sounds like you might have bigger ambitions for it?

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
````diff
@@ -0,0 +1,290 @@
+Be changed to notify users of this ability:
+
+```
+note: compile failed due to 2 errors. You can compile again with `--explain errors` for more information
````
to clarify, this should only happen if in fact there is more information to show?

jonathandturner replied:
Right, with the assumption that in the fullness of time, we have --explain errors cover most, if not all, errors.

nrc commented on text/0000-default-and-expanded-rustc-errors.md:
````diff
@@ -0,0 +1,290 @@
+
+# Stabilization
+
+Currently, these new rust error format is available on nightly using the ```export RUST_NEW_ERROR_FORMAT=true``` environment variable.  Ultimately, this should become the default.  In order to get there, we need to ensure that the new error format is indeed an improvement over the existing format in practice.  
````
nit grammar: "these ... is"

jonathandturner commented:
@nrc 

re:lack of experience with the templated format, are you specifically referring to the style of the text?  In this RFC, we introduce only using the user's code rather than writing our own examples.  We already have a fair bit of experience writing --explain text.  It's true, though, that we haven't implemented this version, so I think it'd be natural to also keep it behind a flag for a cycle, like @alexcrichton suggested.  I can see it maturing at a different rate than the new default errors, so we may decide one or the other needs longer bake time and opting to stabilize one first then the other.

You're right, though, it will need some iteration on the `--explain errors` side style-wise to make sure we're striking the right tone.

nrc commented:
I was referring to the whole thing of templated explains. AIUI, we haven't implemented any version, or have I missed something? I thought the one cycle behind a flag was with the latest version of the extended errors, rather than extended explain, which presumably would need to be behind a flag or something until it stabilised (for some definition).

The tone seems like something that can be evolved separately from the technical components, and it feels like the next RFC will deal with that?

jonathandturner commented:
@nrc - right, I'm just pointing out that we went a little out of order with the revised default errors in that we had it behind a flag before the RFC rather than the other way around.  :)  Yeah, the expanded ones can be behind a flag.

And yes, there will be a follow-up RFC on the tone.  I mention in in this RFC in the last line of the summary.

alexcrichton commented:
🔔 This RFC is now entering its **final comment period** 🔔 

---

My own personal opinion is that I'm super excited to see this in the compiler and the new error format is *great*. Having any functionality behind a flag for at least a cycle makes me comfortable in terms of stability, and this is also an area where the specifics I at least believe are relatively loose as tools would primarily desire a JSON-like interface rather than parse the current format (which is in the works).

jonathandturner commented:
Just to follow-up here, since nrc and I were chatting a bit offline.

Basically, because of the different levels of maturity of the default and the extended errors, we'll be working on both behind a flag.  Because of this different maturity level, both default and extended errors will be behind flag.  Default will align with this rfc, stay behind flag for a cycle, then move to being on by default.  Extended errors will finish experimentation, get a follow-up rfc with implementation details, and this will kick off its final cycle in moving out from under flag.

alexcrichton commented:
@jonathandturner sounds good to me, could you ensure that the text of the RFC reflects that as well?

jonathandturner commented:
@alexcrichton - updated!

brson commented:
I'm happy with this, though I think the `--explain errors` flag is awkward. If not for legacy issues I would expect this to be a single flag (`--explain` :p). I might prefer this to be `--explain-errors`. Frustrating design constraints.

I agree with @nrc that in hindsight it would have been preferable for this RFC to be just about the new error format, and not deal with `--explain` at all. The two features seem orthogonal.

brson commented:
This RFC only specifies 1 feature name, but there are two stabilization cycles here, one for the new errors and one for the new explanations. Can a second feature be added?

brson commented:
Can this RFC address the matter of editors parsing the new text? What's the story for rust-mode specifically? Is this parseable with a regex? The `Drawbacks` section seems to imply this, but the examples show key error information on multiple lines so it's not obvious that a single regex can parse out errors.

jonathandturner commented:
@brson - I'd be happy to explore deprecating --explain.  Since we're already going to have another RFC cycle to move the expanded errors along, perhaps we can explore replacing --explain there?

Not sure what you mean by "This RFC only specifies 1 feature name".  Default errors and expanded errors are both explained here.  Maybe I missed what you meant?

I've reached out to IDE authors, and they're aware of the change.  Some are talking about doing JSON support, some only have minimal support for errors and don't parse, and we're working with those who are parsing the text and need to update their parser (vim specifically).  I think this is okay to be offline with those authors since it'll be different for different editors.  The regex details here are more specific to the regex the editors use.  There are updated techniques, eg @nikomatsakis has updated his emacs install to support the newer errors, so the knowledge there is growing.

Before we turn it on by default, it'd be good to at least have editors moving to support it so that the support can be there when it's turned on in a stable release.

brson commented:
> @brson - I'd be happy to explore deprecating --explain. Since we're already going to have another RFC cycle to move the expanded errors along, perhaps we can explore replacing --explain there?

Yes. Please make it an unresolved question here.

> Not sure what you mean by "This RFC only specifies 1 feature name". Default errors and expanded errors are both explained here. Maybe I missed what you meant?

[This line](https://github.com/rust-lang/rfcs/pull/1644/files#diff-b0c527c5078053b6c1c7cab068e1bab3R1) names a single feature, "default_and_expanded_errors_for_rustc". Presumably that one will apply to the new error format, and be attached to its tracking issue. What name will be attached to the tracking issue for enhanced `--explain`?

> I've reached out to IDE authors, and they're aware of the change. Some are talking about doing JSON support, some only have minimal support for errors and don't parse, and we're working with those who are parsing the text and need to update their parser (vim specifically). I think this is okay to be offline with those authors since it'll be different for different editors. The regex details here are more specific to the regex the editors use. There are updated techniques, eg @nikomatsakis has updated his emacs install to support the newer errors, so the knowledge there is growing.

I don't think this addresses the concern that a single regex can pull out the error information from the text. In this example:

```
+error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
+  --> src/test/compile-fail/borrowck/borrowck-borrow-from-owned-ptr.rs:29:22
```

The word 'error' is one one line, and the file name and number are on another. It's not obvious to me that this can be reliably parsed as a single regex, which I suspect (bot don't know), is a requirement for less capable editors. `rust-mode` is probably not the right barometer since emacs can execute arbitrary lisp to parse buffers. IDEs are not the main concern since they are powerful and can use JSON output - it's lesser editors like e.g. Kate. Unfortunately I don't have a concrete example of an editor that gets just a single regex to parse the error buffer (maybe @nrc or @vadimcn know).

jonathandturner commented:
"The word 'error' is one one line, and the file name and number are on another. It's not obvious to me that this can be reliably parsed as a single regex, which I suspect (bot don't know), is a requirement for less capable editors."  <- this is why we are reaching out to the authors of those editors.  We're also looking at things we can do on our end, like better buffering of the error.

jonathandturner commented:
"What name will be attached to the tracking issue for enhanced --explain?"

Does this need to be named in this RFC?  I'm confused.

alexcrichton commented:
The tools team got a chance to discuss this RFC the other day, and the decision was to merge!

There were some key points brought up while discussing this that the tools team felt needed to be addressed before stabilization, but implementation on an unstable basis seemed fine. Specifically, we were thinking:

* **Editors** - before stabilizing, the new errors need to be *proven* to be parseable in a number of editors. For example we should have a command-line editor like vim/emacs working but we should also have a higher level IDE like IntelliJ working. Note that we're not requiring that every editor under the sun is up to date, just that the basics are proven out and no showstoppers arise. 
* **--explain** - a roughly separate portion of this RFC, we're going to consider it separately for stability from the main error format. That is, we think the new error format will get turned on by default before the `--explain` changes are available in a stable compiler. This will likely also require a further RFC depending on the extent of the changes.

In the meantime, we'll continue to have an opt-in to using the new error format (currently an environment variable) which will allow users to test out the errors, editors to start to catch up, etc.

Thanks again for all the discussion on this RFC everyone, and thanks @jonathandturner for pushing on this! Getting lots of people happy with a stylistic decision is no small feat!

alexcrichton commented:
Tracking issue for default errors: https://github.com/rust-lang/rust/issues/34826
Tracking issue for `--explain`: https://github.com/rust-lang/rust/issues/34827

Oh and one last point --

> Does this need to be named in this RFC? I'm confused.

Ah as part of `--explain` we don't really have a feature name like we do for libraries, so I think it's fine to just create this on the fly like we do with many others
